src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,527 @@
+        kvStoreInsertionPos += 2;
+    }
+
+    private void resizeIndexIfNeeded(int size) {
+        if (size < numSlotsForIndexRebuild) {
+            return;
+        }
+
+        int[] oldIndex = this.index;
+        this.index = new int[2 * index.length];
+
+        for (int i = 0; i < oldIndex.length; i += 2) {
+            int hash = oldIndex[i];
+            int kvPos = oldIndex[i + 1];
+
+            insertIntoIndex(hash, kvPos);
+        }
+        this.numSlotsForIndexRebuild = (int) (this.index.length / 2 * THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD);
+    }
+
+    private void resizeKvStoreIfNeeded() {
+        if (kvStoreInsertionPos >= kvStore.length) {
+            kvStore = Arrays.copyOf(kvStore, 2 * kvStore.length);
+        }
+    }
+
+    private CompactHashStore copy() {
+        return new CompactHashStore(index.clone(), kvStore.clone(), kvStoreInsertionPos, numSlotsForIndexRebuild);
+    }
+
+    private Object deleteKvAndGetV(RubyHash hash, int hashPos, int keyKvPos) {
+        Object deletedValue = kvStore[keyKvPos + 1];
+        index[hashPos + 1] = INDEX_SLOT_DELETED_MARKER;
+        // TODO: Instead of naively nulling out the key-value, which can produce long gaps of nulls in the kvStore,
+        //       See if we can annotate each gap with its length so that iteration code can "jump" over it
+        kvStore[keyKvPos] = null;
+        kvStore[keyKvPos + 1] = null;
+
+        hash.size--;
+        return deletedValue;
+    }
+
+    /** @return the position of the first non-null key from the end of array, or a negative value if all keys are
+     *         null */
+    private int getLastKeyPos() {
+        int lastKeyPos = kvStoreInsertionPos - 2;
+        while (lastKeyPos > 0 && kvStore[lastKeyPos] == null) {
+            lastKeyPos -= 2;
+        }
+        return lastKeyPos;
+    }
+
+    /** @param h a hash code
+     * @param max the length of the index array
+     * @return An even index ranging from 0 to max-1 */
+    private static int getIndexPosFromHash(int h, int max) {
+        return h & (max - 2);
+    }
+
+    // Or -1 if key doesn't exist
+    @GenerateUncached
+    abstract static class GetKeyPosInKVStoreNode extends RubyBaseNode {
+        public abstract int execute(Object key, int hash, int hashPosInIndex, boolean compareByIdentity,
+                int[] index, Object[] kvStore);
+
+        @Specialization
+        protected int getKeyPos(
+                Object key, int hash, int hashPosInIndex, boolean compareByIdentity, int[] index, Object[] kvStore,
+                @Cached CompareHashKeysNode compareHashKeysNode) {
+            if (hashPosInIndex == KEY_NOT_FOUND) {
+                return KEY_NOT_FOUND;
+            }
+            int kvPos = index[hashPosInIndex + 1] - 1;
+            Object otherKey = kvStore[kvPos];
+
+            // identity implies equality
+            if (otherKey == key) {
+                return kvPos;
+            }
```

Comment:
That should be removed, `CompareHashKeysNode` will do the appropriate check.
Notably it's wrong for boxed primitives to use `==`.